myhoard/restore_coordinator.py

```diff
@@ -1135,6 +1135,22 @@ def _check_sql_slave_status(self):
         with self._mysql_cursor() as cursor:
             cursor.execute("SHOW SLAVE STATUS")
             slave_status = cursor.fetchone()
+
+            # Sometimes mysqld can be OOM-killed during backup restoration.
+            # In that case IO and/or SQL thread might not be runnig, so they must be restarted.
+            io_thread_running = slave_status["Slave_IO_Running"]
+            sql_thread_running = slave_status["Slave_SQL_Running"]
+            refetch_slave_status = False
+            if io_thread_running != "Yes":
+                cursor.execute("START SLAVE IO_THREAD")
+                refetch_slave_status = True
+            if sql_thread_running != "Yes":
+                cursor.execute("START SLAVE SQL_THREAD")
```

Comment:
I'd add some backoff logic or max attempts for this. There are cases where starting the thread helps and cases where it doesn't and in the latter it will just make the situation more difficult to assess when the thread is being restarted like crazy. This condition is expected to be very rare and likely a single restart will fix it or then not so perhaps exponential backoff with initial interval of 10 minutes and exponent of 1.5 and max value of 120 minutes would be somewhat reasonable.

Reply:
Thanks for the explanation, will update with the delay/retry logic.